### PR TITLE
Implement RegisterAllocator and ConstantCache

### PR DESCRIPTION
**Pull request by codingpotato**
* Use RegisterAllocator and ConstantCache to emit correct byte code for eval statements, constant expressions and binary expressions.
* Remove useless chunk test cases.

### CODE REVIEW

File: csharp/src/SeedLang/Interpreter/RegisterAllocator.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace SeedLang.Interpreter {
+  // The allocator class to allocate register slots for local and temporary variables.
+  internal class RegisterAllocator {
+    // The maximum allocated registers count. This number is never decreased even when a register is
+    // deallocated.
+    public uint MaxRegisterCount = 0;
```
Comment (wixette) [inline csharp/src/SeedLang/Interpreter/RegisterAllocator.cs:20]: What's the use of `MaxRegisterCount`? If the filed `_registerCount` is pointed to the current boundary of the allocation, what's the reason that we need to get an ever-reached maximum number?

Seems there is no other code using `MaxRegisterCount` in this PR. Is it useful in the future somehow?

File: csharp/src/SeedLang/Interpreter/Chunk.cs [hunk 2]
```diff
@@ -42,16 +45,24 @@ internal class Chunk {
...
     internal uint AddConstant(double number) {
       _constants.Add(new Value(number));
-      return (uint)_constants.Count - 1;
+      return (uint)_constants.Count - 1 + _maxRegisterCount;
+    }
+
+    // Converts the constant id to the index in the constants list.
+    private int IndexOfConstId(uint constId) {
+      Debug.Assert(constId >= _maxRegisterCount && constId - _maxRegisterCount < _constants.Count,
+                   "Constant id is not in the range of the constants list.");
```
Comment (wixette) [inline csharp/src/SeedLang/Interpreter/Chunk.cs:59]:
```suggestion
                   "Constant id is not in the range of the constant list.");
```

File: csharp/src/SeedLang/Interpreter/RegisterAllocator.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    // Current allocated registers count. This number is decreased if a register is deallocated.
+    private uint _registerCount = 0;
+
+    // Allocates a temporary variable and returns the index of the register slot.
+    internal uint AllocateTempVariable() {
```
Comment (wixette) [inline csharp/src/SeedLang/Interpreter/RegisterAllocator.cs:25]: Where or how to ensure the allocation won't exceed the limit of 250 registers?

File: csharp/src/SeedLang/Interpreter/Chunk.cs [hunk 4]
```diff
@@ -19,7 +19,10 @@
 namespace SeedLang.Interpreter {
   // A data structure to hold bytecode and constants generated from the AST tree by the compiler.
   internal class Chunk {
+    // The maximum number of registers that can be allocated in the stack of a chunk.
+    private const uint _maxRegisterCount = 250;
     private readonly List<Instruction> _bytecode = new List<Instruction>();
+    // The constants list to hold all the constants used in this chunk.
```
[the remark this answers is not in the document]
Comment (codingpotato) [reply csharp/src/SeedLang/Interpreter/Chunk.cs:25]: Done

File: csharp/src/SeedLang/Interpreter/Chunk.cs [hunk 5]
```diff
@@ -42,16 +45,24 @@ internal class Chunk {
       _bytecode.Add(new Instruction(opcode, a, bx));
     }
 
+    // Adds a number constant into the constants list and returns the id of the input constant.
```
[the remark this answers is not in the document]
Comment (codingpotato) [reply csharp/src/SeedLang/Interpreter/Chunk.cs:48]: Done

File: csharp/src/SeedLang/Interpreter/Chunk.cs [hunk 6]
```diff
@@ -42,16 +45,24 @@ internal class Chunk {
...
+    //
+    // The returned constant id is the index in the constants list plus the maximum register count.
```
[the remark this answers is not in the document]
Comment (codingpotato) [reply csharp/src/SeedLang/Interpreter/Chunk.cs:50]: Done

File: csharp/src/SeedLang/Interpreter/Chunk.cs [hunk 7]
```diff
@@ -42,16 +45,24 @@ internal class Chunk {
...
     internal uint AddConstant(double number) {
       _constants.Add(new Value(number));
-      return (uint)_constants.Count - 1;
+      return (uint)_constants.Count - 1 + _maxRegisterCount;
+    }
+
+    // Converts the constant id to the index in the constants list.
```
[the remark this answers is not in the document]
Comment (codingpotato) [reply csharp/src/SeedLang/Interpreter/Chunk.cs:56]: Done

File: csharp/src/SeedLang/Interpreter/Chunk.cs (hunk 2 above)
Comment (codingpotato) [reply csharp/src/SeedLang/Interpreter/Chunk.cs:59]: Done

File: csharp/src/SeedLang/Interpreter/ConstantCache.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+
+namespace SeedLang.Interpreter {
+  // A cache class to cache the constant id of constants. It only adds the unique constant into the
+  // constants list of the chunk.
```
[the remark this answers is not in the document]
Comment (codingpotato) [reply csharp/src/SeedLang/Interpreter/ConstantCache.cs:20]: Done

File: csharp/src/SeedLang/Interpreter/ConstantCache.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  internal class ConstantCache {
+    private readonly Chunk _chunk;
+    private readonly Dictionary<double, uint> _numbers = new Dictionary<double, uint>();
+
+    internal ConstantCache(Chunk chunk) {
+      _chunk = chunk;
+    }
+
+    // Returns the id of the given number constant and adds the constant into the constants list of
```
[the remark this answers is not in the document]
Comment (codingpotato) [reply csharp/src/SeedLang/Interpreter/ConstantCache.cs:29]: Done

File: csharp/src/SeedLang/Interpreter/RegisterAllocator.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace SeedLang.Interpreter {
+  // The allocator class to allocate register slots for local and temporary variables.
+  internal class RegisterAllocator {
+    // The maximum allocated registers count. This number is never decreased even when a register is
```
[the remark this answers is not in the document]
Comment (codingpotato) [reply csharp/src/SeedLang/Interpreter/RegisterAllocator.cs:18]: Done

File: csharp/src/SeedLang/Interpreter/RegisterAllocator.cs (hunk 1 above)
Comment (codingpotato) [reply csharp/src/SeedLang/Interpreter/RegisterAllocator.cs:20]: The compiler will copy MaxRegisterCount into Chunk after compilation, which will be used as the allocation size of the registers in the stack during execution. Added code to handle the copy logic. The register allocation logic during execution will be added later.

File: csharp/src/SeedLang/Interpreter/RegisterAllocator.cs (hunk 3 above)
Comment (codingpotato) [reply csharp/src/SeedLang/Interpreter/RegisterAllocator.cs:25]: Added a TODO to throw the exception later.

**Review by wixette — APPROVED** [review]

**Pull request merged**